### PR TITLE
Adding Crowdin integration to l10n scripts.

### PR DESCRIPTION
**Pull request by mkarolin**
Resolves https://github.com/brave/brave-browser/issues/42870

After transitioning to Crowdin, we can remove Transifex-related code in a separate issue/PR.

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: script/lib/l10n/crowdin/common.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    'brave_generated_resources.xml',
+    'brave_components_strings.xml',
+    'brave_extension.json',
+]
+
+
+def should_use_crowdin_for_file(source_string_path, filename):
+    """ Determines if the given file should be handled by Crowdin locally"""
+    name = crowdin_name_from_filename(source_string_path, filename)
+    return name in crowdin_handled_files or name.startswith('greaselion_')
```
Comment (emerick) [inline script/lib/l10n/crowdin/common.py:37]: I doubt we need to support the Greaselion translations any more. cc: @zenparsing

File: build/commands/scripts/commands.js [hunk 2]
```diff
@@ -188,12 +188,16 @@ program
 
 program
   .command('pull_l10n')
+  .option('--service <service>', 'Service to use: Transifex or Crowdin')
+  .option('--channel <channel>', 'Release|Beta|Nightly for Crowdin, Release for Transifex')
   .option('--grd_path <grd_path>', `Relative path to match end of full GRD path, e.g: 'generated_resources.grd'.`)
   .option('--debug', `Dumps downloaded content for one language into TransifexCurrent.txt file in the temp directory.`)
```
Comment (emerick) [inline build/commands/scripts/commands.js:194]: Does this description (and underlying functionality) need to be updated to use the appropriate service name?

File: build/commands/scripts/commands.js [hunk 3]
```diff
@@ -188,12 +188,16 @@ program
...
   .action(pullL10n)
 
 program
   .command('push_l10n')
+  .option('--service <service>', 'Service to use: Transifex or Crowdin')
+  .option('--channel <channel>', 'Release|Beta|Nightly for Crowdin, Release for Transifex')
   .option('--grd_path <grd_path>', `Relative path to match end of full GRD path, e.g: 'generated_resources.grd'.`)
   .option('--with_translations', 'Push local translations. WARNING: this will overwrite translations in Tansifex.')
   .option('--with_missing_translations', 'Push local translations for strings that do not have translations in Transifex.')
```
Comment (emerick) [inline build/commands/scripts/commands.js:203]: Same question as above.

File: script/pull-l10n.py [hunk 4]
```diff
@@ -23,6 +25,14 @@ def parse_args():
     parser = argparse.ArgumentParser(description='Pull strings from Transifex')
```
Comment (emerick) [inline script/pull-l10n.py:25]: or Crowdin?

File: script/push-l10n.py [hunk 5]
```diff
@@ -26,7 +31,15 @@
 
 def parse_args():
     parser = argparse.ArgumentParser(description='Push strings to Transifex')
```
Comment (emerick) [inline script/push-l10n.py:33]: Need to update description.

File: script/lib/l10n/crowdin/common.py (hunk 1 above)
Comment (mkarolin) [reply script/lib/l10n/crowdin/common.py:37]: Oh yeah, nice find. I removed some other greaselion code, but missed this.

Comment (mkarolin) [reply script/lib/l10n/crowdin/common.py]: Fixed in [13140e2](https://github.com/brave/brave-core/pull/27005/commits/13140e28ea7cd53e4056067f06938b892d409c16)

File: build/commands/scripts/commands.js (hunk 2 above)
Comment (mkarolin) [reply build/commands/scripts/commands.js]: Fixed in [13140e2](https://github.com/brave/brave-core/pull/27005/commits/13140e28ea7cd53e4056067f06938b892d409c16)

File: build/commands/scripts/commands.js (hunk 3 above)
Comment (mkarolin) [reply build/commands/scripts/commands.js]: Fixed in [13140e2](https://github.com/brave/brave-core/pull/27005/commits/13140e28ea7cd53e4056067f06938b892d409c16)

File: script/pull-l10n.py (hunk 4 above)
Comment (mkarolin) [reply script/pull-l10n.py]: Fixed in [13140e2](https://github.com/brave/brave-core/pull/27005/commits/13140e28ea7cd53e4056067f06938b892d409c16)

File: script/push-l10n.py (hunk 5 above)
Comment (mkarolin) [reply script/push-l10n.py]: Fixed in [13140e2](https://github.com/brave/brave-core/pull/27005/commits/13140e28ea7cd53e4056067f06938b892d409c16)

File: build/commands/scripts/commands.js [hunk 6]
```diff
@@ -191,16 +191,16 @@ program
...
   .action(pullL10n)
 
 program
   .command('push_l10n')
   .option('--service <service>', 'Service to use: Transifex or Crowdin')
   .option('--channel <channel>', 'Release|Beta|Nightly for Crowdin, Release for Transifex')
   .option('--grd_path <grd_path>', `Relative path to match end of full GRD path, e.g: 'generated_resources.grd'.`)
-  .option('--with_translations', 'Push local translations. WARNING: this will overwrite translations in Tansifex.')
-  .option('--with_missing_translations', 'Push local translations for strings that do not have translations in Transifex.')
+  .option('--with_translations', 'Push local translations. WARNING: this will overwrite translations in Tansifex/Crowdin.')
```
Comment (emerick) [inline build/commands/scripts/commands.js:202]: Nit: Still has the "Tansifex" typo

Comment (mkarolin) [reply build/commands/scripts/commands.js]: Oops, fixed!

File: script/lib/l10n/crowdin/common.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+#
+# Copyright (c) 2024 The Brave Authors. All rights reserved.
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import json
+import re
+import os
+import lxml.etree  # pylint: disable=import-error
```
[the remark this answers is not in the document]
Comment (mkarolin) [reply script/lib/l10n/crowdin/common.py:11]: The xml files we process here don't use DOCTYPE or ENTITY elements.

File: script/lib/l10n/crowdin/api_v2_client_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+        branch_id = self.__get_branch(branch)
+        assert branch_id, (
+            f'Unable to get resource {resource_name} for ' +
+            f'branch {branch} because the branch doesn\'t exist')
+        file_id = self.__get_resource_file(branch_id, resource_name)
+        assert file_id, (
+            f'Unable to get resource {resource_name} for ' +
+            f'branch {branch} because the resource doesn\'t exist')
+        url = self.__get_resource_download_url(file_id)
+        r = requests.get(url)
```
[the remark this answers is not in the document]
Comment (mkarolin) [reply script/lib/l10n/crowdin/api_v2_client_wrapper.py]: Fixed in https://github.com/brave/brave-core/pull/27005/commits/3e089dee3fcf7660a0c853b8f928f319319f56d8

File: script/lib/l10n/crowdin/api_v2_client_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+            assert branch_id, (
+                f'Unable to get {resource_name} l10n for ' +
+                f'branch {branch} because the branch doesn\'t exist')
+            file_id = self.__get_resource_file(branch_id, resource_name)
+            assert file_id, (
+                f'Unable to get {resource_name} l10n for ' +
+                f'branch {branch} because the resource doesn\'t exist')
+            url = self.__get_resource_translation_download_url(
+                file_id, lang_code)
+            r = requests.get(url)
```
[the remark this answers is not in the document]
Comment (mkarolin) [reply script/lib/l10n/crowdin/api_v2_client_wrapper.py]: Fixed in https://github.com/brave/brave-core/pull/27005/commits/3e089dee3fcf7660a0c853b8f928f319319f56d8

**Review by iefremov — APPROVED** [review]

**Review by thypon — APPROVED** [review]

**Pull request merged**